### PR TITLE
Upgrade server and MongoDB python requirements

### PR DESCRIPTION
**Pull request by j053y**
## What changes are proposed in this pull request?

- Upgrade `starlette`, `strawberry`, `mongoengine` and `motor` requirements to the highest version.
  - Use approximate equivalence for `starlette`, `strawberry`, and `motor` to allow for future upgrades.
  - Use hard equivalence for `mongoengine` as it's not updated often.
- Remove version requirement from `pymongo`, instead allowing `mongengine` and `motor` to find the highest common version.
- Allow `deserialize_numpy_array` in `fiftyone.core.utils` to skip de-serialization if already de-serialized. (`pymongo` or `mongoengine` must have changed something with their default serialization as it was coming to the method already de-serialized._)

## How is this patch tested? If it is not, please explain why.

Install with new requirements, run regression tests, and ensure SDK still functions.

### Is this a user-facing change that should be mentioned in the release notes?

-   [x] No. You can skip the rest of this section.
-   [ ] Yes. Give a description of this change to be included in the release
        notes for FiftyOne users.

### What areas of FiftyOne does this PR affect?

-   [ ] App: FiftyOne application changes
-   [ ] Build: Build and test infrastructure changes
-   [x] Core: Core `fiftyone` Python library changes
-   [ ] Documentation: FiftyOne documentation changes
-   [ ] Other

### CODE REVIEW

File: fiftyone/core/utils.py [hunk 1]
```diff
@@ -1285,6 +1285,10 @@ def deserialize_numpy_array(numpy_bytes, ascii=False):
     Returns:
         the numpy array
     """
+
+    if isinstance(numpy_bytes, list):
```
Comment (j053y) [inline fiftyone/core/utils.py:1289]: When I originally upgraded the regression tests were failing (expected `bytes` and not `list`). It seems like the upgraded de-serialized the byes into a list. 

Making this  allowed the regressions to complete without further issue.

**Comment by brimoor** [comment]
@j053y okay, I tested this PR and here's my findings:

## Mongo requirements

**Bad news**: here's the packages that impose `pymongo` requirements:

```
apache-beam==2.42.0 requires pymongo<4.0.0,>=3.8.0
mongoengine==0.24.2 requires pymongo<5.0,>=3.4
motor==3.1.1        requires pymongo<5,>=4.1
motor==2.5.1        requires pymongo<4,>=3.12
```

Our Apache beam utils are "extras", not core. Either way though, the latest `apache-beam` version still doesn't support `pymongo>=4`, so we have to relax to `motor>=2.3` to continue to allow users to install Beam if they wish 😢 

**Good news**: for users that don't care about Apache Beam, I'm confident that we can support `motor v3` and `pymongo v4`. I pushed a different fix for the `deserialize_numpy_array` bug (not sure why this wasn't needed prior to this, tbh) and both regression tests and manual usage of the App/SDK are working fine.

It's also fine to remove `apache-beam` from the extra requirements so that source installers can have `motor v3` and `pymongo v4`.

## App requirements

There are no App tests, so we just have to `launch_app()` and play around. I upgraded to:

```
starlette==0.21.0
strawberry-graphql==0.138.1
```

but was getting errors in the App. So for now I reverted to the previously pinned values. Haven't played around with intermediate versions yet. Perhaps @benjaminpkane can weigh in here?

## Summary

tbh I forget the original impetus for this PR. Do you _need_ `pymongo`/`motor` upgraded? Or `starlette`/`strawberry-graphql`? Or both?

**Comment by j053y** [comment]
@brimoor I need starlette and strawberry.

**Comment by j053y** [comment]
motor 3.x would be great be not as necessary

**Comment by j053y** [comment]
@brimoor and @benjaminpkane I was finally able to get the app running locally. I updated the versions for `strawberry` and `starlette` again and can successfully load and use both the SDK and app.

**Comment by brimoor** [comment]
@j053y so the App works with these for you?

```
starlette==0.20.4
strawberry-graphql==0.138.1
```

I tried upgrading both to latest and it did not work for me with these versions:

```
starlette==0.21.0
strawberry-graphql==0.138.1
```

So I do not think we can use `>=` in the requirements.

But I defer to @benjaminpkane as the final authority on what we can support here.

**Comment by j053y** [comment]
@brimoor Gotcha I changed `==` to   `>=` change after I tested it seems.

The following explicit versions did in fact work for me though.
```
starlette==0.20.4
strawberry-graphql==0.138.1
``` 
Out of curiosity what errors are you getting or is the app just not loading at all for you? I know we're using different OSs, with more information I might be able to narrow in on the problem.

**Comment by brimoor** [comment]
With

```
starlette==0.21.0
strawberry-graphql==0.96.0
```

I get:

```
SyntaxError: Unexpected token 'T', "Traceback "... is not valid JSON
```

With 

```
starlette==0.21.0
strawberry-graphql==0.138.1
```

I get:

```
__init__() takes 1 positional argument but 3 were given
File "/Users/Brian/dev/env/fo/lib/python3.9/site-packages/graphql/execution/execute.py", line 1036, in await_result
return build_response(await result, errors) # type: ignore
File "/Users/Brian/dev/env/fo/lib/python3.9/site-packages/graphql/execution/execute.py", line 451, in get_results
await gather(*(results[field] for field in awaitable_fields)),
File "/Users/Brian/dev/env/fo/lib/python3.9/site-packages/graphql/execution/execute.py", line 535, in await_result
self.handle_field_error(error, return_type)
File "/Users/Brian/dev/env/fo/lib/python3.9/site-packages/graphql/execution/execute.py", line 569, in handle_field_error
raise error
File "/Users/Brian/dev/env/fo/lib/python3.9/site-packages/graphql/execution/execute.py", line 528, in await_result
return_type, field_nodes, info, path, await result
File "/Users/Brian/dev/env/fo/lib/python3.9/site-packages/strawberry/schema/schema_converter.py", line 513, in _async_resolver
return await await_maybe(_get_result(_source, strawberry_info, **kwargs))
File "/Users/Brian/dev/env/fo/lib/python3.9/site-packages/strawberry/utils/await_maybe.py", line 12, in await_maybe
return await value
File "/Users/Brian/dev/fiftyone/fiftyone/server/paginator.py", line 108, in paginate
return await get_items(
File "/Users/Brian/dev/fiftyone/fiftyone/server/paginator.py", line 76, in get_items
edges.append(Edge(node=from_db(doc), cursor=str(_id)))
File "/Users/Brian/dev/fiftyone/fiftyone/server/paginator.py", line 105, in from_db
doc = cls.modifier(doc)
File "/Users/Brian/dev/fiftyone/fiftyone/server/query.py", line 184, in modifier
doc["group_media_types"] = [
File "/Users/Brian/dev/fiftyone/fiftyone/server/query.py", line 185, in <listcomp>
Group(name, media_type)
```

**Comment by j053y** [comment]
@brimoor Weird. With those exact versions it works fine for me. 
```
┍ ⦗22-11-03 10:48:47⦘ josey@mercury: ‹feature/upgrade-requirements›*
╰╼ ~/code/voxel51/fiftyone› pip list | grep starlette 
sse-starlette               0.10.3
starlette                   0.21.0
┍ ⦗22-11-03 10:48:54⦘ josey@mercury: ‹feature/upgrade-requirements›*
╰╼ ~/code/voxel51/fiftyone› pip list | grep strawberry
strawberry-graphql          0.138.1
```
I guess back to the original plan and wait for @benjaminpkane

**Comment by benjaminpkane** [comment]
Strawberry changed their "dataclass" decorator to always mark fields as kwargs. Odd, but things are working now.

**Comment by brimoor** [comment]
@benjaminpkane the App initially loads for me, but when I refresh in browser I get `Not found`.

**Comment by benjaminpkane** [comment]
Ok, static file serving changed. Fixed now.

**Comment by brimoor** [comment]
It sounded like you guys wanted the latest starlette so I bumped it to 0.21.0

File: fiftyone/server/static/404.html [hunk 2]
```diff
@@ -0,0 +1 @@
+
```
Comment (brimoor) [inline fiftyone/server/static/404.html:1]: My git client was having problems pulling this file unless it was non-empty

**Review by brimoor — APPROVED** [review]
LGTM

**Pull request merged**